server/remote_asset/fetch_server/fetch_server.go

```diff
@@ -203,11 +186,70 @@ func (p *FetchServer) FetchDirectory(ctx context.Context, req *rapb.FetchDirecto
+		Hash: checksum,
+		// The digest size is unknown since the client only sends up
+		// the hash. We can look up the size using the Metadata API,
+		// which looks up only using the hash, so the size we pass here
+		// doesn't matter.
+		SizeBytes: 1,
+	}
+	expectedHash := blobDigest.Hash
+	cacheRN := digest.NewResourceName(blobDigest, instanceName, rspb.CacheType_CAS, checksumFunc)
+	log.CtxInfof(ctx, "Looking up %s in cache", blobDigest.Hash)
+
+	// Lookup metadata to get the correct digest size to be returned to
+	// the client.
+	cache := p.env.GetCache()
+	md, err := cache.Metadata(ctx, cacheRN.ToProto())
+	if err != nil {
+		log.CtxInfof(ctx, "FetchServer failed to get metadata for %s: %s", expectedHash, err)
+		return nil
+	}
+	blobDigest.SizeBytes = md.DigestSizeBytes
+
+	// Even though we successfully fetched metadata, we need to renew
+	// the cache entry (using Contains()) to ensure that it doesn't
+	// expire by the time the client requests it from cache.
+	cacheRN = digest.NewResourceName(blobDigest, instanceName, rspb.CacheType_CAS, checksumFunc)
+	exists, err := cache.Contains(ctx, cacheRN.ToProto())
+	if err != nil {
+		log.CtxErrorf(ctx, "Failed to renew %s: %s", digest.String(blobDigest), err)
+		return nil
+	}
+	if !exists {
+		log.CtxInfof(ctx, "Blob %s expired before we could renew it", digest.String(blobDigest))
+		return nil
+	}
+
+	// If the digestFunc is supplied and differ from the checksum sri,
+	// after looking up the cached blob using checksum sri, re-upload
+	// that blob using the requested digestFunc.
+	if checksumFunc != storageFunc {
+		b, err := cache.Get(ctx, cacheRN.ToProto())
+		if err != nil {
+			log.CtxErrorf(ctx, "Failed to get cache reader for %s: %s", digest.String(blobDigest), err)
+			return nil
+		}
+		bsClient := p.env.GetByteStreamClient()
+		storageDigest, err := cachetools.UploadBlob(ctx, bsClient, instanceName, storageFunc, bytes.NewReader(b))
+		if err != nil {
+			log.CtxErrorf(ctx, "Failed to re-upload blob with new digestFunc %s for %s: %s", storageFunc, digest.String(blobDigest), err)
+			return nil
+		}
+		blobDigest = storageDigest
+	}
+
+	log.CtxInfof(ctx, "FetchServer found %s in cache", digest.String(blobDigest))
+	return blobDigest
+}
+
 // mirrorToCache uploads the contents at the given URI to the given cache,
 // returning the digest. The fetched contents are checked against the given
-// expectedSHA256 (if non-empty), and if there is a mismatch then an error is
+// expectedHash (if non-empty), and if there is a mismatch then an error is
 // returned.
-func mirrorToCache(ctx context.Context, bsClient bspb.ByteStreamClient, remoteInstanceName string, httpClient *http.Client, uri, expectedSHA256 string) (*repb.Digest, error) {
+func mirrorToCache(ctx context.Context, bsClient bspb.ByteStreamClient, remoteInstanceName string, httpClient *http.Client, uri string, storageFunc repb.DigestFunction_Value, checksumFunc repb.DigestFunction_Value, checksum string) (*repb.Digest, error) {
```

Comment:
nit: rename `checksum` to `expectedHash` to match the function docs. (Or `expectedChecksum` if you like "checksum" better - I think the "expected" part is useful because this is not the verified checksum, it's the unverified request checksum)

Reply:
Renamed all to `expectedChecksum` to be consistent with the qualifier name.